src/server/saved_objects/migrations/core/elastic_index.ts

```diff
@@ -0,0 +1,284 @@
+
+  if (!err) {
+    return;
+  }
+
+  const exception: any = new Error(err.index.error!.reason);
+  exception.detail = err;
+  throw exception;
+}
+
+/**
+ * Checks to see if the specified index is up to date. It does this by checking
+ * that the index has the expected mappings and by counting
+ * the number of documents that have a property which has migrations defined for it,
+ * but which has not had those migrations applied. We don't want to cache the
+ * results of this function (e.g. in context or somewhere), as it is important that
+ * it performs the check *each* time it is called, rather than memoizing itself,
+ * as this is used to determine if migrations are complete.
+ *
+ * @param {CallCluster} callCluster
+ * @param {string} index
+ * @param {MigrationVersion} migrationVersion - The latest versions of the migrations
+ */
+export async function migrationsUpToDate(
+  callCluster: CallCluster,
+  index: string,
+  migrationVersion: MigrationVersion
+): Promise<boolean> {
+  const indexInfo = await fetchInfo(callCluster, index);
+
+  if (!_.get(indexInfo, 'mappings.doc.properties.migrationVersion')) {
+    return false;
+  }
+
+  // If no migrations are actually defined, we're up to date!
+  if (Object.keys(migrationVersion).length <= 0) {
+    return true;
+  }
+
+  const { count } = await callCluster('count', {
+    body: {
+      query: {
+        bool: {
+          should: Object.entries(migrationVersion).map(([type, latestVersion]) => ({
+            bool: {
+              must: [
+                { exists: { field: type } },
+                { bool: { must_not: { term: { [`migrationVersion.${type}`]: latestVersion } } } },
+              ],
+            },
+          })),
+        },
+      },
+    },
+    index,
+    type: 'doc',
+  });
+
+  return count === 0;
+}
+
+/**
+ * Applies the specified mappings to the index.
+ *
+ * @param {CallCluster} callCluster
+ * @param {string} index
+ * @param {IndexMapping} mappings
+ */
+export function putMappings(callCluster: CallCluster, index: string, mappings: IndexMapping) {
+  return callCluster('indices.putMapping', { body: mappings.doc, index, type: 'doc' });
+}
+
+export async function createIndex(
+  callCluster: CallCluster,
+  index: string,
+  mappings?: IndexMapping
+) {
+  await callCluster('indices.create', { body: { mappings }, index });
+}
+
+export async function deleteIndex(callCluster: CallCluster, index: string) {
+  await callCluster('indices.delete', { index });
+}
+
+/**
+ * Converts the specified index to an alias.
+ *
+ * @param {CallCluster} callCluster
+ * @param {FullIndexInfo} info
+ * @param {string} alias
+ */
+export async function convertToAlias(callCluster: CallCluster, info: FullIndexInfo, alias: string) {
+  await callCluster('indices.create', {
+    body: { mappings: info.mappings },
+    index: info.indexName,
+  });
+
+  await callCluster('reindex', {
+    body: {
+      dest: { index: info.indexName },
+      source: { index: alias },
```

Comment:
This really threw me for a loop - alias here is actually not the alias, but the originating Kibana index. Mind adding some information to the function description to cover this?

Reply:
... Yeah. This is a case of tricky naming. `alias` is currently an index, but we want it to be an alias, so this is converting it to an alias. `info` is an IndexInfo object that contains the name of the reindex target... I'll add this info to the doc comment. I'd love to rename everything to make this clearer, if you can think of better names.